### ISSUE TITLE
1.11.1: build fails with `[-Werror=odr]` when LTO is used

### ISSUE THREAD

**Issue by kloczek** [openucx/ucx#7329]
### Describe the bug
Looks like LTO exposes some issues.
Build fails with `[-Werror=odr]` when LTO is used.
I'm using gcc 11.2.1.
```console
make[3]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen'
make[3]: Entering directory '/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest'
/bin/sh ../../libtool  --tag=CXX   --mode=link /usr/bin/g++ -O3 -g -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -fno-tree-vectorize -DGTEST_UCM_HOOK_LIB_DIR="\"/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen/.libs\""  -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none  -no-install -Wl,-dynamic-list-data  -Wl,-z,relro -Wl,--as-needed -Wl,--gc-sections -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto=auto -flto-partition=none -fuse-linker-plugin -o gtest common/gtest-gtest-all.o common/gtest-main.o common/gtest-test_gtest_cmn.o common/gtest-mem_buffer.o common/gtest-test_helpers.o common/gtest-test_obj_size.o common/gtest-test_watchdog.o common/gtest-test_perf.o common/gtest-test.o ucm/gtest-malloc_hook.o uct/gtest-test_amo.o uct/gtest-test_amo_add_xor.o uct/gtest-test_amo_and_or.o uct/gtest-test_amo_cswap.o uct/gtest-test_amo_fadd_fxor.o uct/gtest-test_amo_fand_for.o uct/gtest-test_amo_swap.o uct/gtest-test_event.o uct/gtest-test_fence.o uct/gtest-test_flush.o uct/gtest-test_many2one_am.o uct/gtest-test_md.o uct/gtest-test_mm.o uct/gtest-test_mem.o uct/gtest-test_p2p_am.o uct/gtest-test_p2p_err.o uct/gtest-test_p2p_mix.o uct/gtest-test_p2p_rma.o uct/gtest-test_pending.o uct/gtest-test_progress.o uct/gtest-test_uct_ep.o uct/gtest-test_uct_perf.o uct/v2/gtest-test_uct_query.o uct/gtest-test_zcopy_comp.o uct/gtest-uct_p2p_test.o uct/gtest-uct_test.o uct/gtest-test_stats.o ucs/gtest-test_event_set.o ucs/gtest-test_stats_filter.o uct/gtest-test_peer_failure.o uct/gtest-test_tag.o uct/tcp/gtest-test_tcp.o ucp/gtest-test_ucp_am.o ucp/gtest-test_ucp_ep.o ucp/gtest-test_ucp_stream.o ucp/gtest-test_ucp_peer_failure.o ucp/gtest-test_ucp_atomic.o ucp/gtest-test_ucp_dt.o ucp/gtest-test_ucp_tls.o ucp/gtest-test_ucp_memheap.o ucp/gtest-test_ucp_mmap.o ucp/gtest-test_ucp_mem_type.o ucp/gtest-test_ucp_perf.o ucp/gtest-test_ucp_proto.o ucp/gtest-test_ucp_rma.o ucp/gtest-test_ucp_rma_mt.o ucp/gtest-test_ucp_tag_cancel.o ucp/gtest-test_ucp_tag_match.o ucp/gtest-test_ucp_tag_offload.o ucp/gtest-test_ucp_tag_mt.o ucp/gtest-test_ucp_tag_perf.o ucp/gtest-test_ucp_tag_probe.o ucp/gtest-test_ucp_tag_xfer.o ucp/gtest-test_ucp_tag_mem_type.o ucp/gtest-test_ucp_tag.o ucp/gtest-test_ucp_context.o ucp/gtest-test_ucp_worker.o ucp/gtest-test_ucp_wireup.o ucp/gtest-test_ucp_request.o ucp/gtest-test_ucp_wakeup.o ucp/gtest-test_ucp_fence.o ucp/gtest-test_ucp_sockaddr.o ucp/gtest-ucp_test.o ucp/gtest-ucp_datatype.o ucs/gtest-test_algorithm.o ucs/gtest-test_arbiter.o ucs/gtest-test_async.o ucs/gtest-test_bitmap.o ucs/gtest-test_callbackq.o ucs/gtest-test_class.o ucs/gtest-test_config.o ucs/gtest-test_conn_match.o ucs/gtest-test_datatype.o ucs/gtest-test_bitops.o ucs/gtest-test_debug.o ucs/gtest-test_memtrack.o ucs/gtest-test_math.o ucs/gtest-test_mpmc.o ucs/gtest-test_mpool.o ucs/gtest-test_pgtable.o ucs/gtest-test_profile.o ucs/gtest-test_rcache.o ucs/gtest-test_memtype_cache.o ucs/gtest-test_stats.o ucs/gtest-test_strided_alloc.o ucs/gtest-test_string.o ucs/gtest-test_sys.o ucs/gtest-test_topo.o ucs/gtest-test_sock.o ucs/gtest-test_time.o ucs/gtest-test_twheel.o ucs/gtest-test_frag_list.o ucs/gtest-test_type.o ucs/gtest-test_log.o ucs/gtest-test_iov.o ucs/gtest-test_vfs.o ucs/arch/gtest-test_x86_64.o uct/ib/gtest-test_ib.o uct/ib/gtest-test_ib_md.o uct/ib/gtest-test_cq_moderation.o uct/ib/gtest-test_ib_xfer.o uct/ib/gtest-test_ib_pkey.o uct/ib/gtest-test_ib_event.o uct/ib/gtest-test_devx.o uct/ib/gtest-ud_base.o uct/ib/gtest-test_ud.o uct/ib/gtest-test_ud_timer.o uct/ib/gtest-test_ud_pending.o uct/ib/gtest-test_ud_ds.o uct/ib/gtest-test_rc.o uct/ib/gtest-test_dc.o uct/ib/gtest-test_sockaddr.o   ../../src/ucs/libucs.la ../../src/ucs/vfs/sock/libucs_vfs_sock.la ../../src/uct/libuct.la ../../src/ucm/libucm.la ../../src/ucp/libucp.la ../../src/tools/perf/lib/libucxperf.la -fopenmp  -libverbs ../../src/uct/ib/libuct_ib.la   -lrt -lrt  -lbfd -ldl
libtool: link: /usr/bin/g++ -O3 -g -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -fno-tree-vectorize -DGTEST_UCM_HOOK_LIB_DIR=\"/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen/.libs\" -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -Wl,-dynamic-list-data -Wl,-z -Wl,relro -Wl,--as-needed -Wl,--gc-sections -Wl,-z -Wl,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto=auto -flto-partition=none -fuse-linker-plugin -o gtest common/gtest-gtest-all.o common/gtest-main.o common/gtest-test_gtest_cmn.o common/gtest-mem_buffer.o common/gtest-test_helpers.o common/gtest-test_obj_size.o common/gtest-test_watchdog.o common/gtest-test_perf.o common/gtest-test.o ucm/gtest-malloc_hook.o uct/gtest-test_amo.o uct/gtest-test_amo_add_xor.o uct/gtest-test_amo_and_or.o uct/gtest-test_amo_cswap.o uct/gtest-test_amo_fadd_fxor.o uct/gtest-test_amo_fand_for.o uct/gtest-test_amo_swap.o uct/gtest-test_event.o uct/gtest-test_fence.o uct/gtest-test_flush.o uct/gtest-test_many2one_am.o uct/gtest-test_md.o uct/gtest-test_mm.o uct/gtest-test_mem.o uct/gtest-test_p2p_am.o uct/gtest-test_p2p_err.o uct/gtest-test_p2p_mix.o uct/gtest-test_p2p_rma.o uct/gtest-test_pending.o uct/gtest-test_progress.o uct/gtest-test_uct_ep.o uct/gtest-test_uct_perf.o uct/v2/gtest-test_uct_query.o uct/gtest-test_zcopy_comp.o uct/gtest-uct_p2p_test.o uct/gtest-uct_test.o uct/gtest-test_stats.o ucs/gtest-test_event_set.o ucs/gtest-test_stats_filter.o uct/gtest-test_peer_failure.o uct/gtest-test_tag.o uct/tcp/gtest-test_tcp.o ucp/gtest-test_ucp_am.o ucp/gtest-test_ucp_ep.o ucp/gtest-test_ucp_stream.o ucp/gtest-test_ucp_peer_failure.o ucp/gtest-test_ucp_atomic.o ucp/gtest-test_ucp_dt.o ucp/gtest-test_ucp_tls.o ucp/gtest-test_ucp_memheap.o ucp/gtest-test_ucp_mmap.o ucp/gtest-test_ucp_mem_type.o ucp/gtest-test_ucp_perf.o ucp/gtest-test_ucp_proto.o ucp/gtest-test_ucp_rma.o ucp/gtest-test_ucp_rma_mt.o ucp/gtest-test_ucp_tag_cancel.o ucp/gtest-test_ucp_tag_match.o ucp/gtest-test_ucp_tag_offload.o ucp/gtest-test_ucp_tag_mt.o ucp/gtest-test_ucp_tag_perf.o ucp/gtest-test_ucp_tag_probe.o ucp/gtest-test_ucp_tag_xfer.o ucp/gtest-test_ucp_tag_mem_type.o ucp/gtest-test_ucp_tag.o ucp/gtest-test_ucp_context.o ucp/gtest-test_ucp_worker.o ucp/gtest-test_ucp_wireup.o ucp/gtest-test_ucp_request.o ucp/gtest-test_ucp_wakeup.o ucp/gtest-test_ucp_fence.o ucp/gtest-test_ucp_sockaddr.o ucp/gtest-ucp_test.o ucp/gtest-ucp_datatype.o ucs/gtest-test_algorithm.o ucs/gtest-test_arbiter.o ucs/gtest-test_async.o ucs/gtest-test_bitmap.o ucs/gtest-test_callbackq.o ucs/gtest-test_class.o ucs/gtest-test_config.o ucs/gtest-test_conn_match.o ucs/gtest-test_datatype.o ucs/gtest-test_bitops.o ucs/gtest-test_debug.o ucs/gtest-test_memtrack.o ucs/gtest-test_math.o ucs/gtest-test_mpmc.o ucs/gtest-test_mpool.o ucs/gtest-test_pgtable.o ucs/gtest-test_profile.o ucs/gtest-test_rcache.o ucs/gtest-test_memtype_cache.o ucs/gtest-test_stats.o ucs/gtest-test_strided_alloc.o ucs/gtest-test_string.o ucs/gtest-test_sys.o ucs/gtest-test_topo.o ucs/gtest-test_sock.o ucs/gtest-test_time.o ucs/gtest-test_twheel.o ucs/gtest-test_frag_list.o ucs/gtest-test_type.o ucs/gtest-test_log.o ucs/gtest-test_iov.o ucs/gtest-test_vfs.o ucs/arch/gtest-test_x86_64.o uct/ib/gtest-test_ib.o uct/ib/gtest-test_ib_md.o uct/ib/gtest-test_cq_moderation.o uct/ib/gtest-test_ib_xfer.o uct/ib/gtest-test_ib_pkey.o uct/ib/gtest-test_ib_event.o uct/ib/gtest-test_devx.o uct/ib/gtest-ud_base.o uct/ib/gtest-test_ud.o uct/ib/gtest-test_ud_timer.o uct/ib/gtest-test_ud_pending.o uct/ib/gtest-test_ud_ds.o uct/ib/gtest-test_rc.o uct/ib/gtest-test_dc.o uct/ib/gtest-test_sockaddr.o -fopenmp  ../../src/ucs/.libs/libucs.so ../../src/ucs/vfs/sock/.libs/libucs_vfs_sock.a ../../src/uct/.libs/libuct.so ../../src/ucm/.libs/libucm.so ../../src/ucp/.libs/libucp.so ../../src/tools/perf/lib/.libs/libucxperf.a /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucp/.libs/libucp.so ../../src/uct/ib/.libs/libuct_ib.so -libverbs -lmlx5 /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/uct/.libs/libuct.so /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/.libs/libucs.so -lnuma -lm /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucm/.libs/libucm.so -lrt -lbfd -ldl -fopenmp -Wl,-rpath -Wl,/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/.libs -Wl,-rpath -Wl,/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/uct/.libs -Wl,-rpath -Wl,/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucm/.libs -Wl,-rpath -Wl,/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucp/.libs -Wl,-rpath -Wl,/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/uct/ib/.libs -Wl,-rpath -Wl,/usr/lib64/ucx
ucs/test_async.cc:20: error: type 'struct base' violates the C++ One Definition Rule [-Werror=odr]
   20 | class base {
      |
ucs/test_class.cc:16: note: a type with different virtual table pointers is defined in another translation unit
   16 | typedef struct base {
      |
lto1: all warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/g++ returned 1 exit status
```

**Comment by kloczek** [comment]
Just checked and when LTO is disabled build fails but with other issues
```console
make[4]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen'
make[3]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen'
make[3]: Entering directory '/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest'
/usr/bin/g++ -DHAVE_CONFIG_H -I. -I../..  -DCPU_FLAGS="" -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1 -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src -DTOP_SRCDIR=\"../..\" -I../../src -I../../test -I../../src -I../../test/gtest  -fopenmp     -O3 -g -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -fno-tree-vectorize -DGTEST_UCM_HOOK_LIB_DIR="\"/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen/.libs\""  -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -c -o ucs/gtest-test_bitops.o `test -f 'ucs/test_bitops.cc' || echo './'`ucs/test_bitops.cc
/usr/bin/g++ -DHAVE_CONFIG_H -I. -I../..  -DCPU_FLAGS="" -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1 -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src -DTOP_SRCDIR=\"../..\" -I../../src -I../../test -I../../src -I../../test/gtest  -fopenmp     -O3 -g -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -fno-tree-vectorize -DGTEST_UCM_HOOK_LIB_DIR="\"/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen/.libs\""  -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -c -o uct/ib/gtest-test_dc.o `test -f 'uct/ib/test_dc.cc' || echo './'`uct/ib/test_dc.cc
/usr/bin/g++ -DHAVE_CONFIG_H -I. -I../..  -DCPU_FLAGS="" -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1 -I/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src -DTOP_SRCDIR=\"../..\" -I../../src -I../../test -I../../src -I../../test/gtest  -fopenmp     -O3 -g -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -fno-tree-vectorize -DGTEST_UCM_HOOK_LIB_DIR="\"/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/test/gtest/ucm/test_dlopen/.libs\""  -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -c -o uct/ib/gtest-test_sockaddr.o `test -f 'uct/ib/test_sockaddr.cc' || echo './'`uct/ib/test_sockaddr.cc
In file included from /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/sys/math.h:17,
                 from /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/sys/string.h:12,
                 from ./common/test_helpers.h:19,
                 from ./common/test.h:19,
                 from ucs/test_bitops.cc:6:
/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/arch/bitops.h: In member function ‘virtual void test_bitops_is_equal_Test::test_body()’:
/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/arch/bitops.h:186:14: error: array subscript 20 is outside array bounds of ‘uint8_t [20]’ {aka ‘unsigned char [20]’} [-Werror=array-bounds]
  186 |     return ((*((uint8_t*)ptr1 + length) & ~UCS_MASK(8 - remainder)) ==
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~
ucs/test_bitops.cc:87:13: note: while referencing ‘buffer1’
   87 |     uint8_t buffer1[20] = {0};
      |             ^~~~~~~
In file included from /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/sys/math.h:17,
                 from /home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/sys/string.h:12,
                 from ./common/test_helpers.h:19,
                 from ./common/test.h:19,
                 from ucs/test_bitops.cc:6:
/home/tkloczko/rpmbuild/BUILD/ucx-1.11.1/src/ucs/arch/bitops.h:187:14: error: array subscript 20 is outside array bounds of ‘uint8_t [20]’ {aka ‘unsigned char [20]’} [-Werror=array-bounds]
  187 |             (*((uint8_t*)ptr2 + length) & ~UCS_MASK(8 - remainder)));
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~
ucs/test_bitops.cc:88:13: note: while referencing ‘buffer2’
   88 |     uint8_t buffer2[20] = {0};
      |             ^~~~~~~
cc1plus: all warnings being treated as errors
make[3]: *** [Makefile:2698: ucs/gtest-test_bitops.o] Error 1
```

**Comment by kloczek** [comment]
Just in case I'm setting source code tree with:
```console
+ ./configure --build=x86_64-redhat-linux-gnu --host=x86_64-redhat-linux-gnu --program-prefix= --disable-dependency-tracking --disable-silent-rules --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc --datadir=/usr/share --includedir=/usr/include --libdir=/usr/lib64 --libexecdir=/usr/libexec --localstatedir=/var --runstatedir=/run --sharedstatedir=/var/lib --mandir=/usr/share/man --infodir=/usr/share/info --disable-assertions --disable-debug --disable-logging --disable-compiler-opt --disable-optimizations --disable-params-check --disable-static --enable-cma --enable-gtest --with-rdmacm --with-verbs --with-ib-hw-tm --without-cuda --without-gdrcopy --without-java --without-knem --without-rocm --without-ugni --without-xpmem
```
